pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -0,0 +1,346 @@
+        equivalent to U up to a global phase.
+    """
+    # Check unitarity
+    if not math.allclose(math.dot(U, math.T(math.conj(U))), math.eye(4), atol=1e-7):
+        raise ValueError("Operator must be unitary.")
+
+    # Compute the determinant
+    det = math.linalg.det(U)
+
+    # Convert to SU(4) if it's not close to 1
+    if not math.allclose(det, 1.0):
+        exp_angle = -1j * math.cast_like(math.angle(det), 1j) / 4
+        U = math.cast_like(U, det) * qml.math.exp(exp_angle)
+
+    return U
+
+
+def _select_rotation_angles(U):
+    r"""Choose the rotation angles of RZ, RY in the two-qubit decomposition.
+    They are chosen as per Proposition V.1 in quant-ph/0308033 and are based
+    on the phases of the eigenvalues of :math:`E^\dagger \gamma(U) E`, where
+
+    .. math::
+
+        \gamma(U) = (E^\dag U E) (E^\dag U E)^T.
+    """
+
+    gammaU = qml.math.linalg.multi_dot([Edag, U, E, Et, qml.math.T(U), qml.math.T(Edag)])
+    evs = qml.math.linalg.eigvals(gammaU)
+
+    # The rotation angles can be computed as follows (any three eigenvalues can be used)
+    x, y, z = qml.math.angle(evs[0]), qml.math.angle(evs[1]), qml.math.angle(evs[2])
+
+    # Compute functions of the eigenvalues; there are different options in v1
+    # vs. v3 of the paper, I'm not entirely sure why.
+    alpha = (x + y) / 2
+    beta = (x + z) / 2
+    delta = (z + y) / 2
+
+    return alpha, beta, delta
+
+
+def _su2su2_to_tensor_products(U):
+    """Given a matrix U = A \otimes B in SU(2) x SU(2), extract the two SU(2)
+    operations A and B.
+
+    This process has been described in detail in the Appendix of Coffey & Deiotte
+    https://link.springer.com/article/10.1007/s11128-009-0156-3
+    """
+
+    # First, write A = [[a1, a2], [-a2*, a1*]], which we can do for any SU(2) element.
+    # Then, A \otimes B = [[a1 B, a2 B], [-a2*B, a1*B]] = [[C1, C2], [C3, C4]]
+    # where the Ci are 2x2 matrices.
+    C1 = U[0:2, 0:2]
+    C2 = U[0:2, 2:4]
+    C3 = U[2:4, 0:2]
+    C4 = U[2:4, 2:4]
+
+    # From the definition of A \otimes B, C1 C4^\dag = a1^2 I, so we can extract a1
+    C14 = qml.math.dot(C1, qml.math.conj(qml.math.T(C4)))
+    a1 = qml.math.sqrt(C14[0, 0])
+
+    # Similarly, -C2 C3^\dag = a2^2 I, so we can extract a2
+    C23 = qml.math.dot(C2, qml.math.conj(qml.math.T(C3)))
+    a2 = qml.math.sqrt(-C23[0, 0])
+
+    # This gets us a1, a2 up to a sign. To resolve the sign, ensure that
+    # C1 C2^dag = a1 a2* I
+    C12 = qml.math.dot(C1, qml.math.conj(qml.math.T(C2)))
+
+    if not qml.math.isclose(a1 * np.conj(a2), C12[0, 0]):
+        a2 *= -1
+
+    # Construct A
+    A = qml.math.stack([[a1, a2], [-qml.math.conj(a2), qml.math.conj(a1)]])
+
+    # Next, extract B. Can do from any of the C, just need to be careful in
+    # case one of the elements of A is 0.
+    if not qml.math.isclose(A[0, 0], 0.0, atol=1e-8):
+        B = C1 / A[0, 0]
+    else:
+        B = C2 / A[0, 1]
```

Comment:
This is something that would cause issues if we support tracing 🤔 

Could it be replaced with usage of `qml.math.where`? Minor though, happy to keep this.